src/QueuePerThreadPool.c

```diff
@@ -272,21 +276,32 @@ static void claim_work(QPTPoolThreadData_t *tw) {
 static void dump_queue_size_stats(QPTPool_t *ctx, QPTPoolThreadData_t *tw) {
     pthread_mutex_lock(&ctx->mutex);
     pthread_mutex_lock(&print_mutex);
-    tw->waiting.size = tw->claimed.size;
 
-    struct timespec now;
-    clock_gettime(CLOCK_MONOTONIC, &now);
-    fprintf(stderr, "qptpool_size %" PRIu64 " ", since_epoch(&now) - epoch);
+    if (ctx->incomplete) {
```

Comment:
Oops. I should have said: `if (ctx->incomplete > ctx->queue_size_highwater)`.

Reply:
After experimenting with this, I ended up questioning the overall approach I took in this PR.

I tried approximately what you suggested:
```
    if (ctx->incomplete > ctx->queue_size_highwater) {
        if (sum > ctx->queue_size_highwater)
            ctx->queue_size_highwater = sum;

         // print out stats ...
    }
```

but realized that `sum` and `ctx->incomplete` end up being very different from each other most of the time.

`incomplete` will usually be higher than `sum` because any threads that are currently processing work (so between `claim_work(tw);` and `ctx->incomplete -= work_count;` in `worker_function()`) will have an amount of outstanding work that is not reflected by summing up the `waiting` and `deferred` queue sizes. We could also include the `claimed` queue size in `sum`, but this would still tend to be less than `incomplete` since `incomplete` isn't updated until after the entire `claimed` queue is processed.

On the other hand, `incomplete` could be lower than the sum of the `waiting` and `deferred` queue sizes if this function runs concurrently with `QPTPool_enqueue()` and `waiting.size` is checked after enqueuing but before incrementing `incomplete`. 

And that brings up another potential issue, which is that I think there is a data race here because `QPTPool_enqueue()` writes to some `QPTPoolThreadData->waiting.size` (while protected by `QPTPoolThreadData->mutex`), but `dump_queue_size_stats()` reads each `QPTPoolThreadData->waiting.size` without any mutex protection, so it can race with that write. To do this correctly, `dump_queue_size_stats()` should probably acquire each `QPTPoolThreadData->mutex` before reading `waiting.size`...

But in any event, I'm not sure what the best approach is here to actually accomplish what I wanted, which is just to track the highwater mark for outstanding work items. I could either:
- track incomplete directly, with the risk of occasionally overestimating or underestimating the actual amount of outstanding work
- sum all three of the queues, but with a need to acquire more locks to do it completely race-free

do you have any thoughts?

Reply:
I think I was wrong about using `ctx->incomplete`, and `dump_queue_size_stats()` should just print instantaneous queue sizes and the sum.